### PR TITLE
Use codecov

### PR DESCRIPTION
**Pull request by jolting**
LICENSE is BSD
Based on this repo:
https://github.com/RWTH-HPC/CMake-codecov

Note this changes the behavior of make test. The old behavior is make test_legacy.

---

I acknowledge to have:
* Read the [`CONTRIBUTING`](https://github.com/MRPT/mrpt/blob/master/.github/CONTRIBUTING.md) page
* Updated [`doc/doxygen-pages/changeLog_doc.h`](https://github.com/MRPT/mrpt/blob/master/doc/doxygen-pages/changeLog_doc.h) to describe these changes (if applicable)
* Updated [`AUTHORS`](https://github.com/MRPT/mrpt/blob/master/AUTHORS) (if applicable)

(Notify: @MRPT/owners )

### CODE REVIEW

**Comment by jolting** [comment]
Here is what it looks like:
https://codecov.io/gh/jolting/mrpt/branch/master

**Comment by jlblancoc** [comment]
Looks great! 👍 

I've just tried this new setup under CMake+MSVC, and it complains code coverage is not supported with MSVC2017 (?). Anyway, we can live with that. 

The point is, the target "test" is now "test_legacy", and I can see a new CMake target "RUN_TESTS", which tried to run all examples without building them first (!). Perhaps the behavior for Windows is different than with Makefile generators? Will check it out later... 

In any case, I'm unconfortable with this line: 
https://github.com/MRPT/mrpt/pull/642/files#diff-7728197db21ae89029c6a9aff43f1e81R104

Unless dependencies are well defined, `make tests_build_all test` might try to run the tests without building them first? Also, if deps are correct, why not directly skipping the target tests_build_all? I think I'm missing something :-)

**Comment by jolting** [comment]
ctest takes over the `make test` rule, but doesn't build the targets. Perhaps ctest `make test` rule can have target dependencies. I justs couldn't figure it out.

It's based on this:
https://github.com/RWTH-HPC/CMake-codecov/blob/master/.travis.yml

Maybe it's better just to copy and paste this:
```
  - cmake .. -DENABLE_COVERAGE=On && make && make test
  - make gcov
  - make lcov
```

File: travis/travis_main.sh [hunk 1]
```diff
@@ -93,15 +93,20 @@ function test ()
...
     -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+    -DENABLE_COVERAGE=On \
     $EXTRA_CMAKE_ARGS
 
   # Remove gdb use for coverage test reports.
   # Use `test_gdb` to show stack traces of failing unit tests.
 #  if command_exists gdb ; then
 #    make test_gdb
 #  else
+    make tests_build_all
```
Comment (jolting) [inline travis/travis_main.sh]: It does each step independently now.

**Comment by jolting** [comment]
Hmm added the `make lcov`, but it didn't work. Maybe lcov isn't installed in the container.

**Comment by jolting** [comment]
The issue:
https://gitlab.kitware.com/cmake/cmake/issues/8774

**Comment by jlblancoc** [comment]
I think I don't understand what do [the percentage figures here](https://codecov.io/gh/jolting/mrpt/branch/master) actually mean? 
Is really 80% the actual code coverage of *all* code under `libs`, when running all unit tests? I wish it was, but I think the number was pretty smaller when I did some cov tests in the past (?).

**Pull request merged**